### ISSUE TITLE
Better exceptions

### ISSUE THREAD

**Issue by halfpastfouram** [dompdf/dompdf#2316]
The exceptions thrown by this library are not really specific in some cases. One such case is the infamous `Box property calculation requires containing block width on element` exception. It doesn't explain where, why or what is happening.

I would like to propose a better standard for exceptions to help developers determine what they should do to avoid that exception.

Exceptions should be categorized:

1. **ParseErrorException**: The library was unable to parse the provided content. The exception should note what HTML element on what line in the content could not be parsed (DOMDocument node has all the info). This exception should also betriggered when CSS could not be parsed in `Dompdf\Css\Style` and `Dompdf\Css\Stylesheet`.
2. **InvalidArgumentException**: An argument provided to a library method is invalid. Examples of the exceptions that should be converted to this one:
   - `FontNotFoundException`
   - "Child not found in this frame" exception.
   - "Remote file requested, but remote file download is disabled." exception.
3. **RenderException**: An error occurred during render time, typically an internal error.
4. **CacheException**: Anything that went wrong with cache.
5. **FileSystemError**: A file on the filesystem could not be found, read or permission was denied. 

The new **ParseErrorException** would provide users with detailed information about the problem. An example:

```
Box property calculation requires containing block width on line 102: <div class="my-div" style="position: static">
```

Thes simple addition of the line number (`$frame->get_node()->getLineNo()`) and the first line of the HTML node the problem occurred on: `strtok($frame->get_node()->ownerDocument->saveXml($frame->get_node()), "\n")` will save developers a lot of time they would otherwise need to debug the problem.

**Comment by bsweeney** [comment]
I think any work we did related to this would probably fall under #960?

**Comment by halfpastfouram** [comment]
I don't think so. The type of exception thrown and implementing a PSR-3 logger are two different things in my opinion. Changing the type of exceptions makes it easier for anyone to catch and handle specific situations differently. Also, most  applications may already implement a logger that catches exceptions and logs them. So this issue may provide some quick and easy wins on that part.
